test/functional/tests.py

```diff
@@ -1,14 +1,52 @@
 # _*_ coding: utf-8 -*-
 import os
-import shutil
 import json
+import shutil
+import socket
+import sys
 import unittest
+import logging
 
 import ipfsApi
 
 
+__is_available = NotImplemented
+def is_available():
+    """
+    Return whether the IPFS daemon is reachable or not
+    """
+    global __is_available
+    
+    if not isinstance(__is_available, bool):
+        s = socket.socket()
+        try:
+            s.connect((ipfsApi.default_host, ipfsApi.default_port))
+        except IOError:
+            __is_available = False
+        else:
+            __is_available = True
+        finally:
+            s.close()
+    
+    return __is_available
+
+
+def skipIfOffline():
+    if is_available():
+        return lambda func: func
```

Comment:
Decorators (in Python) are actually just pretty syntax around the magic `__call__(*args, **kwargs)` method.

Therefor:

```
@skipIfOffline()
def test_somehting():
    …
```

is the same as:

```
def test_something():
   …
test_something = skipIfOffline()(test_something)
```

were the `skipIfOffline()` part invokes the `skipIfOffline` function and (temporarily) stores the result. The result returned by `skipIfOffline` is in also a callable, that then gets invoked with the actual method as parameter. This second callable will then return *another* callable, that is later invoked as actual unit test by the test harness.

Definitely ask for clarification if this is hard to grasp for you (layers, layers, LAYERS). :wink: